events/handler.go

```diff
@@ -0,0 +1,238 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package events
+
+import (
+	"sort"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/cadvisor/utils/oomparser"
+)
+
+// EventManager is implemented by Events. It provides three main ways to monitor
+// events
+type EventManager interface {
+	// Watch() provides a channel to which EventManager types are streamed as
+	// they are detected, but before being fed to the channel may be screened
+	// by parameter defined in the passed *Request object
+	Watch(eventChannel chan *Event, request *Request) error
+	// GetEvents() returns a slice of all events detected that have passed
+	// the *Request object parameters to the caller
+	GetEvents(request *Request) ([]*Event, error)
+	// AddEvent allows the caller to add an event to an EventManager
+	// instantiation
+	AddEvent(e *Event) error
+}
+
+// Events  holds a slice of *Event objects with a potential field
+// that caps the number of events held.  It is an implementation of the
+// EventManager interface
+type events struct {
+	eventlist     []*Event
+	maxEventsHeld int
+}
+
+// Event contains information general to events such as the time at which they
+// occurred, their specific type, and the actual event. Event types are
+// differentiated by the EventType field of Event.
+type Event struct {
+	ContainerName string
+	TimeOfEvent   time.Time
+	EventType     EventCategory
+	EventData     EventDataInterface
+}
+
+// Request holds a set of parameters by which Event objects may be screened.
+// The caller may want events that occurred within a specific timeframe
+// or of a certain type, which may be specified in the *Request object
+// they pass to an EventManager function
+type Request struct {
+	StartTime   time.Time
+	EndTime     time.Time
+	EventType   map[EventCategory]bool
+	LastNEvents int
+}
+
+// EventCategory is an enumerated type which lists the categories under which
+// events may fall. The Event field EventType is populated by this enum.
+type EventCategory int
+
+const (
+	OomInstanceCat EventCategory = iota
+	ContainerCreation
+	ContainerDeletion
+)
+
+// a general interface which populates the Event field EventData. The actual
+// object, such as an OomInstance, is set as an Event's EventData
+type EventDataInterface interface {
+}
+
+// returns a pointer to an initialized Events object
+func NewEventManager() *events {
+	return &events{
+		eventlist: []*Event{},
+	}
+}
+
+// returns a pointer to an initialized Request object
+func NewRequest() *Request {
+	return &Request{
+		StartTime:   time.Time{},
+		EndTime:     time.Time{},
+		EventType:   map[EventCategory]bool{},
```

Comment:
WDYT of having an EventType be a slice instead of a map? Then we won't need a New for it.